File: lmmx/tools/executor.py
"""Ferramentas de execucao de comandos com suporte a sudo e confirmacao do usuario."""

import subprocess
import shlex
from ..utils.console import confirm_sudo_command, confirm_command, print_command_output

# definicoes de ferramentas para chamada de funcao do groq
EXECUTOR_TOOLS = [
    {
        "type": "function",
        "function": {
            "name": "run_command",
            "description": "Executa um comando shell. Para comandos que modificam o sistema, use run_sudo_command.",
            "parameters": {
                "type": "object",
                "properties": {
                    "command": {
                        "type": "string",
                        "description": "O comando a ser executado (sem sudo)"
                    },
                    "require_confirmation": {
                        "type": "boolean",
                        "description": "Se deve pedir confirmação do usuário antes de executar",
                        "default": True
                    }
                },
                "required": ["command"]
            }
        }
    },
    {
        "type": "function",
        "function": {
            "name": "run_sudo_command",
            "description": "Executa um comando com privilégios sudo. SEMPRE requer confirmação do usuário.",
            "parameters": {
                "type": "object",
                "properties": {
                    "command": {
                        "type": "string",
                        "description": "O comando a ser executado com sudo (sem incluir 'sudo' no comando)"
                    }
                },
                "required": ["command"]
            }
        }
    }
]


def run_command(command: str, require_confirmation: bool = True) -> str:
    """Executa um comando shell."""
    try:
        # verificacao de seguranca - rejeita comandos obviamente perigosos
        dangerous_patterns = ["rm -rf /", "rm -rf /*", ":(){", "mkfs", "dd if="]
        for pattern in dangerous_patterns:
            if pattern in command:
                return f"⛔ comando bloqueado por seguranca: {command}"
        
        # pede confirmacao se necessario
        if require_confirmation:
            if not confirm_command(command):
                return "❌ comando cancelado pelo usuario."
        
        # executa o comando
        result = subprocess.run(
            command,
            shell=True,
            capture_output=True,
            text=True,
            timeout=120
        )
        
        output = result.stdout.strip()
        error = result.stderr.strip()
        
        if result.returncode == 0:
            if output:
                return f"✅ comando executado com sucesso:\n{output}"
            else:
                return "✅ comando executado com sucesso (sem saida)."
        else:
            if error:
                return f"❌ erro (codigo {result.returncode}):\n{error}"
            else:
                return f"❌ comando falhou com codigo {result.returncode}"
    
    except subprocess.TimeoutExpired:
        return "⏱️ erro: comando demorou mais de 2 minutos e foi cancelado."
    except Exception as e:
        return f"❌ erro ao executar comando: {str(e)}"


def run_sudo_command(command: str) -> str:
    """Executa um comando com privilegios sudo."""
    try:
        # remove sudo se ja estiver no comando para evitar "sudo sudo"
        if command.strip().startswith("sudo "):
            command = command.strip()[5:]  # remove prefixo "sudo "
        
        # verificacao de seguranca - rejeita comandos obviamente perigosos
        dangerous_patterns = ["rm -rf /", "rm -rf /*", ":(){", "mkfs.", "dd if=/dev/zero", "chmod -R 777 /"]
        for pattern in dangerous_patterns:
            if pattern in command:
                return f"⛔ comando bloqueado por seguranca: sudo {command}"
        
        # sempre requer confirmacao para comandos sudo
        full_command = f"sudo {command}"
        if not confirm_sudo_command(full_command):
            return "❌ comando sudo cancelado pelo usuario."
        
        # executa com sudo
        result = subprocess.run(
            full_command,
            shell=True,
            capture_output=True,
            text=True,
            timeout=300  # 5 minutos para comandos sudo (instalacao, etc.)
        )
        
        output = result.stdout.strip()
        error = result.stderr.strip()
        
        if result.returncode == 0:
            if output:
                return f"✅ comando sudo executado com sucesso:\n{output}"
            else:
                return "✅ comando sudo executado com sucesso."
        else:
            if error:
                return f"❌ erro sudo (codigo {result.returncode}):\n{error}"
            else:
                return f"❌ comando sudo falhou com codigo {result.returncode}"
    
    except subprocess.TimeoutExpired:
        return "⏱️ erro: comando sudo demorou mais de 5 minutos e foi cancelado."
    except Exception as e:
        return f"❌ erro ao executar comando sudo: {str(e)}"


def execute_executor_tool(tool_name: str, arguments: dict) -> str:
    """Executa uma ferramenta de execucao."""
    tools_map = {
        "run_command": run_command,
        "run_sudo_command": run_sudo_command,
    }
    
    if tool_name not in tools_map:
        return f"erro: tool '{tool_name}' nao encontrada."
    
    return tools_map[tool_name](**arguments)

File: src/tools/executor.ts
// implementacoes de ferramentas de executor - execucao de comando

import { exec } from 'child_process';
import { promisify } from 'util';

const execAsync = promisify(exec);

// padroes perigosos para bloqueio de seguranca
const DANGEROUS_PATTERNS = [
    'rm -rf /',
    'rm -rf /*',
    ':(){',
    'mkfs.',
    'dd if=/dev/zero',
    'chmod -R 777 /',
];

/**
 * verifica se comando e perigoso
 */
function isDangerous(command: string): boolean {
    return DANGEROUS_PATTERNS.some(pattern => command.includes(pattern));
}

/**
 * executa um comando shell
 */
export async function runCommand(command: string): Promise<string> {
    try {
        // verificacao de seguranca
        if (isDangerous(command)) {
            return `⛔ comando bloqueado por seguranca: ${command}`;
        }

        const { stdout, stderr } = await execAsync(command, {
            timeout: 120000, // 2 minutos
            maxBuffer: 1024 * 1024, // 1MB
        });

        const output = stdout.trim();
        const error = stderr.trim();

        if (error && !output) {
            return `❌ erro:\n${error}`;
        }

        if (output) {
            return `✅ comando executado com sucesso:\n${output}`;
        }

        return '✅ comando executado com sucesso (sem saida).';
    } catch (error: any) {
        if (error.killed) {
            return '⏱️ erro: comando demorou mais de 2 minutos e foi cancelado.';
        }
        return `❌ erro ao executar comando: ${error.message}`;
    }
}

/**
 * executa um comando com privilegios sudo
 */
export async function runSudoCommand(command: string): Promise<string> {
    try {
        // remove sudo se ja estiver no comando
        let cleanCommand = command.trim();
        if (cleanCommand.startsWith('sudo ')) {
            cleanCommand = cleanCommand.slice(5);
        }

        // verificacao de seguranca
        if (isDangerous(cleanCommand)) {
            return `⛔ comando bloqueado por seguranca: sudo ${cleanCommand}`;
        }

        const fullCommand = `sudo ${cleanCommand}`;

        const { stdout, stderr } = await execAsync(fullCommand, {
            timeout: 300000, // 5 minutos
            maxBuffer: 1024 * 1024 * 5, // 5MB
        });

        const output = stdout.trim();
        const error = stderr.trim();

        if (error && !output) {
            return `❌ erro sudo:\n${error}`;
        }

        if (output) {
            return `✅ comando sudo executado com sucesso:\n${output}`;
        }

        return '✅ comando sudo executado com sucesso.';
    } catch (error: any) {
        if (error.killed) {
            return '⏱️ erro: comando sudo demorou mais de 5 minutos e foi cancelado.';
        }
        return `❌ erro ao executar comando sudo: ${error.message}`;
    }
}

/**
 * executa uma ferramenta de executor
 */
export async function executeExecutorTool(toolName: string, args: any): Promise<string> {
    switch (toolName) {
        case 'run_command':
            return await runCommand(args.command);
        case 'run_sudo_command':
            return await runSudoCommand(args.command);
        default:
            return `erro: tool '${toolName}' nao encontrada.`;
    }
}
